crates/bevy_ecs/src/storage/thin_array_ptr.rs

```diff
@@ -0,0 +1,297 @@
+        }
+        arr
+    }
+
+    /// Allocate memory for the array, this should only be used if not previous allocation has been made (capacity = 0)
+    /// The caller should update their saved `capacity` value to reflect the fact that it was changed
+    ///
+    /// # Panics
+    /// - Panics if the new capacity overflows `usize`
+    pub fn alloc(&mut self, capacity: NonZeroUsize) {
+        self.set_capacity(capacity.get());
+        if size_of::<T>() != 0 {
+            let new_layout = Layout::array::<T>(capacity.get())
+                .expect("layout should be valid (arithmetic overflow)");
+            // SAFETY:
+            // - layout has non-zero size, `capacity` > 0, `size` > 0 (`size_of::<T>() != 0`)
+            self.data = NonNull::new(unsafe { alloc(new_layout) })
+                .unwrap_or_else(|| handle_alloc_error(new_layout))
+                .cast();
+        }
+    }
+
+    /// Reallocate memory for the array, this should only be used if a previous allocation for this array has been made (capacity > 0).
+    ///
+    /// # Panics
+    /// - Panics if the new capacity overflows `usize`
+    ///
+    /// # Safety
+    /// - the current capacity is indeed greater than 0
+    /// -   The caller should update their saved `capacity` value to reflect the fact that it was changed
+    pub unsafe fn realloc(&mut self, current_capacity: NonZeroUsize, new_capacity: NonZeroUsize) {
+        #[cfg(debug_assertions)]
+        assert_eq!(self.capacity, current_capacity.into());
+        self.set_capacity(new_capacity.get());
+        if size_of::<T>() != 0 {
+            let new_layout =
+                Layout::array::<T>(new_capacity.get()).expect("overflow while allocating memory");
+            // SAFETY:
+            // - ptr was be allocated via this allocator
+            // - the layout of the array is the same as `Layout::array::<T>(current_capacity)`
+            // - the size of `T` is non 0, and `new_capacity` > 0
+            // - "new_size, when rounded up to the nearest multiple of layout.align(), must not overflow (i.e., the rounded value must be less than usize::MAX)",
+            // since the item size is always a multiple of its align, the rounding cannot happen
+            // here and the overflow is handled in `Layout::array`
+            self.data = NonNull::new(unsafe {
+                realloc(
+                    self.data.cast().as_ptr(),
+                    // We can use `unwrap_unchecked` because this is the Layout of the current allocation, it must be valid
+                    Layout::array::<T>(current_capacity.get()).debug_checked_unwrap(),
+                    new_layout.size(),
+                )
+            })
+            .unwrap_or_else(|| handle_alloc_error(new_layout))
+            .cast();
+        }
+    }
+
+    /// Initializes the value at `index` to `value`. This function does not do any bounds checking.
+    ///
+    /// # Safety
+    /// index must be in bounds (`index` <= `len`)
```

Comment:
Lots of references to len which doesn't exist and none to capacity which seems more important here

Reply:
The owner is expected to keep track of the length

Reply:
While that is true there is plenty of documentation that avoids referring to a non-existent variable such as "the length of the ThinColumn" also I don't know from a safety standpoint there is anything wrong with initializing arbitrary members of the ThinColumn as long as you don't exceed capacity. Mostly for this one it felt odd to skip over capacity which certainly has to be respected for safety.